src/audio/detect_test.c

```diff
@@ -0,0 +1,481 @@
+	uint32_t count = frames; /**< Assmuming single channel */
+	uint32_t sample;
+
+	/* synthetic load */
+	if (cd->config)
+		idelay(cd->config->load_mips * 1000000);
+
+	/* compare with the last sample from previous period */
+	if (!cd->detected && abs(cd->prev_sample - src[0]) >=
+	    DETECT_TEST_SPIKE_THRESHOLD) {
+		detect_test_notify(dev);
+		cd->detected = 1;
+	}
+
+	/* perform detection within current period */
+	for (sample = 1; sample < count; ++sample) {
+		if (!cd->detected &&
+		    abs(src[sample - 1] - src[sample]) >=
+		    DETECT_TEST_SPIKE_THRESHOLD) {
+			detect_test_notify(dev);
+			cd->detected = 1;
+		}
+	}
+
+	/* remember last sample from the current period */
+	cd->prev_sample = src[count - 1];
+}
+
+static int free_mem_load(struct comp_data *cd)
+{
+	if (!cd) {
+		trace_keyword_error("free_mem_load() error: invalid cd");
+		return -EINVAL;
+	}
+
+	if (cd->load_memory) {
+		rfree(cd->load_memory);
+		cd->load_memory = NULL;
+	}
+
+	return 0;
+}
+
+static int alloc_mem_load(struct comp_data *cd, uint32_t size)
+{
+	if (!cd) {
+		trace_keyword_error("alloc_mem_load() error: invalid cd");
+		return -EINVAL;
+	}
+
+	free_mem_load(cd);
+
+	if (!size)
+		return 0;
+
+	cd->load_memory = rballoc(RZONE_BUFFER, SOF_MEM_CAPS_RAM, size);
+
+	if (!cd->load_memory) {
+		trace_keyword_error("alloc_mem_load() alloc failed");
+		return -ENOMEM;
+	}
+
+	bzero(cd->load_memory, size);
+
+	return 0;
+}
+
+static void detect_test_free_parameters(struct comp_data *cd)
+{
+	rfree(cd->config);
+	cd->config = NULL;
+}
+
+static struct comp_dev *test_keyword_new(struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct sof_ipc_comp_process *keyword;
+	struct sof_ipc_comp_process *ipc_keyword =
+		(struct sof_ipc_comp_process *)comp;
+	struct comp_data *cd;
+
+	trace_keyword("test_keyword_new()");
+
+	if (IPC_IS_SIZE_INVALID(ipc_keyword->config)) {
+		IPC_SIZE_ERROR_TRACE(TRACE_CLASS_KEYWORD, ipc_keyword->config);
+		return NULL;
+	}
+
+	dev = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+		      COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+
+	keyword = (struct sof_ipc_comp_process *)&dev->comp;
+	memcpy(keyword, ipc_keyword, sizeof(struct sof_ipc_comp_process));
+
+	cd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*cd));
+
+	if (!cd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	/* using default processing function */
+	cd->detect_func = default_detect_test;
+
+	comp_set_drvdata(dev, cd);
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void test_keyword_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	trace_keyword("test_keyword_free()");
+
+	detect_test_free_parameters(cd);
+	free_mem_load(cd);
+	rfree(cd);
+	rfree(dev);
+}
+
+/* set component audio stream parameters */
+static int test_keyword_params(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct sof_ipc_comp_config *config = COMP_GET_CONFIG(dev);
+
+	dev->params.frame_fmt = config->frame_fmt;
```

Comment:
@slawblauciak I think we should use not use config params but use runtime PCM params here. So this can be removed.